### PR TITLE
Fips

### PR DESCRIPTION
**Pull request by v-nisidh**
Add FIPS Support.

### CODE REVIEW

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java [hunk 1]
```diff
@@ -194,6 +194,7 @@ static String getResource(String key)
 				{"R_packetSizePropertyDescription", "The network packet size used to communicate with SQL Server."},
 				{"R_encryptPropertyDescription", "Determines if Secure Sockets Layer (SSL) encryption should be used between the client and the server."},
 				{"R_trustServerCertificatePropertyDescription", "Determines if the driver should validate the SQL Server Secure Sockets Layer (SSL) certificate."},
+				{"R_trustStoreTypePropertyDescription", "Type of trust store type like JKS / PKCS12 or nay FIPS implementation KeyStore Type. etc."},
```
Comment (v-mabarw) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java]: Are these the allowable types or are there others? http://docs.oracle.com/javase/8/docs/technotes/guides/security/StandardNames.html#KeyStore

Perhaps we should simply say 'Type of trust store, such as JKS or PKCS12.'

File: src/main/java/com/microsoft/sqlserver/jdbc/StringUtils.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+		// Hiding constructor.
+	}
+
+	/**
+	 * Checks if String is null 
+	 * @param charSequence {@link CharSequence} Can provide null
+	 * @return {@link Boolean} if provided char sequence is null or empty / blank
+	 * @since 6.1.2
+	 */
+	public static boolean isEmpty(final CharSequence charSequence) {
```
Comment (v-mabarw) [inline src/main/java/com/microsoft/sqlserver/jdbc/StringUtils.java]: Is it worth it to add an entire class for the isEmpty implementation - how much use do we expect to get out of this in the future?

File: src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java [hunk 3]
```diff
@@ -1823,6 +1843,50 @@ void enableSSL(String host, int port) throws SQLServerException
...
+			//Here different FIPS provider supports different KeyStore type along with different JVM Implementation. 
+			isValidTrustStoreType = !StringUtils.isEmpty(trustStoreType);
+			isTrustServerCertificate = con.trustServerCertificate();
+
+			if (isEncryptOn && isValidTrustStoreType && !isTrustServerCertificate) {
+				isFipsEnabled = true;
+			} else {
+				//If FIPS is not enabled issue warning and fall-back to non-FIPS mode.
+				StringBuilder sb = new StringBuilder();
+				sb.append("Could not switch to FIPS mode due to above settings: ");
```
Comment (v-mabarw) [inline src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java]: We cannot build strings directly in the implementation - it needs to be localizable to languages other than English.  Can we add a parameterized string resource and use that (you can see what is done for the error messages, for example)?

File: src/test/java/com/microsoft/sqlserver/jdbc/fips/FipsEnvTest.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+	 */
+	@Test
+	public void testFIPSOnOracle() throws Exception {
+		assumeTrue(ORACLE_JVM.equals(currentJVM), "Aborting test: As this is not Oracle Env. ");
+
+		assumeTrue("FIPS".equals(Utils.getConfiguredProperty("FIPS_ENV")), "Aborting test case as FIPS is not enabled. "); 
+		
+		assertTrue(isFIPS("SunJSSE"),"FIPS Should be enabled");
+		
+//		assertTrue(isFIPS("BCFIPS"),"FIPS Should be enabled");
```
Comment (v-mabarw) [inline src/test/java/com/microsoft/sqlserver/jdbc/fips/FipsEnvTest.java]: What's this one for (is it needed still)?

File: src/test/java/com/microsoft/sqlserver/jdbc/fips/FipsEnvTest.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+		assumingThat("BCFIPS".equals(Utils.getConfiguredProperty("FIPS_ENV")),
+				() -> assertAll("All FIPS", () -> assertTrue(isFIPS("SunJSSE"), "FIPS should be Enabled."), () -> assertTrue(isFIPS("BCFIPS"), "Testing")));
+		
+		assumingThat("FIPS".equals(Utils.getConfiguredProperty("FIPS_ENV")), ()-> assertTrue(isFIPS("SunJSSE"), "FIPS Should be enabled"));
+		*/
+	}
+
+	/**
+	 * It will test
```
Comment (v-mabarw) [inline src/test/java/com/microsoft/sqlserver/jdbc/fips/FipsEnvTest.java]: Finish comment :)

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java (hunk 1 above)
Comment (v-nisidh) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java]: BCFIPS supports both PKCS12 & BCFKS. 

Ref: https://www.bouncycastle.org/fips/BCFipsDescription-20150501.pdf [2.4.8 FIPS Compliant Key Store]

File: src/main/java/com/microsoft/sqlserver/jdbc/StringUtils.java (hunk 2 above)
Comment (v-nisidh) [inline src/main/java/com/microsoft/sqlserver/jdbc/StringUtils.java]: In future we can add more functions if we are going to use this. As per need basis. This is kind of common utilities.  In future, we can add functions like add padding, masking etc. I am seeing this class as place holder for all common / global functions related to String operations.

File: src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java (hunk 3 above)
Comment (v-nisidh) [inline src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java]: I will add localized warning message in coming sprint.

**Pull request merged**